%\VignetteIndexEntry{R package for Intra-Sample Heterogeneity Scores}
%\VignetteKeywords{methylation, RnBeads}

\documentclass{article}

\usepackage[colorlinks=T,linkcolor=blue]{hyperref}
\usepackage{ebgaramond}

\title{ISH - R package for Intra-Sample Heterogeneity Scores}

\author{Michael Scherer}

\begin{document}
\SweaveOpts{concordance=TRUE}

\maketitle

\section{Introduction}

This vignette desribes the functionalities included in the ISH R package. We describe the Intra-Sample Heterogeneity Scores FDRP, qFDRP, PDR, Epipolymorphism and Entropy. The package is able to compute each of those scores from bisulfite sequencing data. Input should be a bam-file that contains reads that have been aligned to a reference genome. While PDR, qFDRP and FDRP are independent of the used mapping tools, Epipolymorphism and Entropy require the reads to be aligned with bismark. Here, we only discuss how to use the package. A detailed description of each of the scores can be found in the corresponding publications.

\section{Installation}

The package is available from GitHub and can be installed by the following command, given that the \href{https://cran.r-project.org/web/packages/devtools/index.html}{devtools} package is installed in your R installation:
<<>>=
devtools::install_github("schmic05/ISH_package")
library(ISH)
@

You can test if the installed version is functioning by employing one of the examples in the package:

<<eval=FALSE>>=
#' cannot be executed on all machines, since ~7GB of RAM are required
qfdrp <- ish.run.example()
@

\section{Computing ISH scores}
\subsection{FDRP, qFDRP and PDR}
FDRP, qFDRP and PDR do not require any additional tools or scripts and can be computed directly from your \textit{bam} file. In this case, the \texttt{score} argument of the \texttt{compute.score} functions needs to be one of "fdrp", "qfdrp" or "pdr". You need to specify the CpG sites for which the scores should be computed in either of two forms: \texttt{GRanges} or \texttt{RnBSet}.
\begin{enumerate}
    \item[GRanges] This object should contain the positions of the CpGs for which analysis is to be conducted. The \texttt{GRanges} object should contain a single entry for each CpG and only have length 1 for each of the entries. Then you can either run \texttt{compute.score.GRanges} directly or call the generic function \texttt{compute.score}.
<<>>=
example.bam <- system.file(file.path("extData","small_example.bam"),package="ISH")
example.GRanges <- GRanges(Rle(rep("chr2",10)),IRanges(start = c(2298361,2298554,2298732,2298743,2298787,2298792,2298827,2298884,
                                                                 2298915,2298921),end=c(2298361,2298554,2298732,2298743,2298787,
                                                                                        2298792,2298827,2298884,2298915,2298921)+1))
pdr <- compute.score(bam.file=example.bam,example.GRanges,score="pdr")
@
    \item[RnBSet] In addition to \texttt{GRanges} objects, the ISH package support \texttt{RnBSet} objects as input. Here, the annotation is inferred from the object's annotation with the addition of only selecting those sites that have a coverage higher than \textit{coverage.threhold} in the \texttt{RnBSet} object. For more details on how to set options for analysis, see \autoref{option_setting}.
<<fig=TRUE>>=
example.rnb.set <- system.file(file.path("extData","small_rnbSet.zip"),package="ISH")
example.rnb.set <- load.rnb.set(example.rnb.set)
set.option(coverage.threshold = 10)
fdrp <- rnb.calculate.fdrp(example.rnb.set,example.bam)
to.plot <- data.frame(qFDRP=qfdrp$qFDRP,FDRP=fdrp$FDRP)
to.plot <- melt(to.plot)
#plot <- ggplot(to.plot,aes(x=value,y=..density..,color=variable))+geom_density()+theme_bw()
#plot
@
\end{enumerate}

\subsection{MHL}
In contrast to the scores above, MHL requires a working version of perl installed on your machine. For Linux, this should in general be \texttt{/usr/bin/perl}, which is per default set in this package. In case you are using MacOS (why we do not support Windows is argued in \autoref{windows_problems}), you first need to specify the option \textit{perl.path}. Furthermore, a working version of \texttt{samtools} is required by the programs that compute MHL.
<<eval=FALSE>>=
set.option(perl.path = "/usr/bin/perl")
set.option(samtools.path = "/usr/bin/samtools")
mhl <- compute.score.GRanges(bam.file = example.bam, range = example.GRanges, score="mhl")
@

\subsection{Epipolymorphism and Entropy}
Epipolymorphism and Entropy calculation depends on the methclone software (\url{https://code.google.com/archive/p/methclone/}) to compute epiallele counts and then uses R functions to compute the final scores. Thus, you need a working version of the software installed on your machine and need to tell the package where this is located. Similar to above, an option (\textit{methclone.path}) is resposible for the configuration.
\section{Advanced Configuration}
\subsection{Option settings}
\label{option_setting}

\subsection{Windows troubleshooting}
\label{windows_problems}

\end{document}
